server/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java

```diff
@@ -105,70 +96,33 @@ public static RestResponse buildTextPlainResponse(Table table, RestChannel chann
             return new RestResponse(RestStatus.OK, RestResponse.TEXT_CONTENT_TYPE, BytesArray.EMPTY);
         }
 
-        return new RestResponse(RestStatus.OK, new ChunkedRestResponseBody() {
-
-            private boolean needsHeader = verbose;
-            private final Iterator<Integer> rowIterator = rowOrder.iterator();
-
-            private RecyclerBytesStreamOutput currentOutput;
-            private final Writer writer = new OutputStreamWriter(new OutputStream() {
-                @Override
-                public void write(int b) throws IOException {
-                    assert currentOutput != null;
-                    currentOutput.write(b);
-                }
-
-                @Override
-                public void write(byte[] b, int off, int len) throws IOException {
-                    assert currentOutput != null;
-                    currentOutput.write(b, off, len);
-                }
-
-                @Override
-                public void flush() {
-                    assert currentOutput != null;
-                    currentOutput.flush();
-                }
-
-                @Override
-                public void close() {
-                    assert currentOutput != null;
-                    currentOutput.flush();
-                }
-            }, StandardCharsets.UTF_8);
-
-            @Override
-            public boolean isDone() {
-                return needsHeader == false && rowIterator.hasNext() == false;
-            }
-
-            @Override
-            public ReleasableBytesReference encodeChunk(int sizeHint, Recycler<BytesRef> recycler) throws IOException {
-                try {
-                    assert currentOutput == null;
-                    currentOutput = new RecyclerBytesStreamOutput(recycler);
-
-                    if (needsHeader) {
-                        needsHeader = false;
-                        for (int col = 0; col < headers.size(); col++) {
-                            DisplayHeader header = headers.get(col);
-                            boolean isLastColumn = col == lastHeader;
-                            pad(
-                                new Table.Cell(header.display, table.findHeaderByName(header.name)),
-                                width[col],
-                                request,
-                                writer,
-                                isLastColumn
-                            );
-                            if (isLastColumn == false) {
-                                writer.append(" ");
+        return new RestResponse(
+            RestStatus.OK,
+            ChunkedRestResponseBody.fromTextChunks(
+                RestResponse.TEXT_CONTENT_TYPE,
+                Iterators.concat(
+                    // optional header
+                    Iterators.single(writer -> {
```

Comment:
I really like the refactoring, separates the mechanics of writing chunks from generating a table very well and code is more readable than before.
This is the only thing that bugs me a little: the if inside the `single` iterator, so that if it is false that iterator is useless. IMO it would be better to have it outside and skip the concat if verbose is false, but it may lead to less "elegant" code. But it's a very minor thing.

Reply:
Thanks @ldematte. This refactoring also opens up a few more places where we can use chunked-encoding too.

In practice there's no problem with returning an occasional no-op chunk here: we keep on consuming chunks until we've filled the whole buffer anyway, so this won't make any difference to how the output appears on the wire. I did try moving the conditional out a level or two but IMO that was a less readable way to express what we're trying to do. Moreover this matches most closely the implementation from earlier versions:

https://github.com/elastic/elasticsearch/blob/d94e6959cb6306603fd92f48d447528996c07583/server/src/main/java/org/elasticsearch/rest/action/cat/RestTable.java#L87-L97

Reply:
Yep that was one option I considered. I don't like it because you kind of have to describe `bodyChunks` first, even though it goes out after the header. Of course you can extract a variable for the header chunk too, but the more variables you extract the further it gets from the equivalent unchunked code.

The other option is to say `verbose ? Iterator.single(...) : Collections.emptyIterator()`, noting that `Iterators.concat` skips over empty iterators automatically. That at least lets you write it in the right order, but still I didn't prefer it.